WIP: generate Warthog model files from templates

dzhelezov opened the pull request:
This is a WIP PR (built on top of https://github.com/Joystream/joystream/pull/641) for https://github.com/Joystream/joystream/issues/673
All the key features are already available.

TODOs:
- [x]  Replace hardcoded paths and namings with calls to warthog's Config
- [x]  Refactor: unify FTS and Model files rendering, extract interfaces, separate rendering and path resolution
- [x] Test: Cover with unittests key entity relations and types

dzhelezov committed:
replace private field access with the getter

dzhelezov committed:
add tests to tsconfig

dzhelezov committed:
fix test build errors

dzhelezov committed:
auto-fix linter errors

dzhelezov committed:
source cli commands from lib/src

dzhelezov committed:
fix debug import to avoid linting errors

dzhelezov committed:
add undefined guard

dzhelezov committed:
add @types/node v 12.12.30

dzhelezov committed:
fix sample schema

dzhelezov committed:
linter fix

dzhelezov committed:
remove lib befor build

dzhelezov committed:
replace yarn call with a direct api call

dzhelezov committed:
switch off debug

dzhelezov committed:
fix lint errors

dzhelezov committed:
add model, service and resolver templates

dzhelezov committed:
add first iteration of the cli-less mustache template renderer and supporting utils

dzhelezov committed:
add a helper method to infer the field DB type

dzhelezov committed:
replace warthog the cli call with the renderer

dzhelezov committed:
rename renderers->generators, ModelCodeGenerator -> WarthogModelBuilder

dzhelezov committed:
refactor and fix tests

dzhelezov committed:
add testing for the pre-push husky hook

dzhelezov committed:
refactor Generate -> Render

dzhelezov committed:
fix FTSQueryGenerator test

dzhelezov committed:
code cleanup

dzhelezov committed:
refactor: replace one-line methods with statements

dzhelezov committed:
refactor: replace one-line methods with statements

dzhelezov committed:
Merge branch 'query_node_no_cli' of github.com:dzhelezov/joystream into query_node_no_cli

metmirr commented on query-node/substrate-query-framework/cli/src/templates/entities/model.ts.mst:
```diff
@@ -0,0 +1,59 @@
+  {{#has.int}}IntField,{{/has.int}}
+  {{#has.numeric}}NumericField,{{/has.numeric}}
+  {{#has.json}}JSONField,{{/has.json}}
+  {{#has.bytes}}JSONField,{{/has.bytes}}
```
Should be `BytesField`

dzhelezov replied:
Fixed + added some unit tests

metmirr commented:
Very impressive work!

Except for relationships (`mto`, `otm`...) it works well with https://github.com/Joystream/joystream/issues/525

There are naming and typescript types problems with related fields. For example with this schema

```graphql
type Author @entity {
  posts: [Post!]
}

type Post @entity {
  title: String
  author: Author!
}
```

column definition for `posts` field will be following:

```typescript
@OneToMany(() => ---, (posts: Posts) => posts.posts)
posts?: Posts[];
```

actually it should be:

```typescript
@OneToMany(() => Post, (post: Post) => post.author)
posts?: Post[];
```

dzhelezov committed:
add method for adding fields one by one

dzhelezov committed:
fix import in the template and add tests for it

dzhelezov committed:
add model generation util from a string

dzhelezov committed:
fix otm and mto reference types

dzhelezov committed:
add tests for mto and otm types

dzhelezov commented:
> Very impressive work!
> 
> Except for relationships (`mto`, `otm`...) it works well with #525
> 
> There are naming and typescript types problems with related fields. For example with this schema
> 
> ```graphql
> type Author @entity {
>   posts: [Post!]
> }
> 
> type Post @entity {
>   title: String
>   author: Author!
> }
> ```
> 
> column definition for `posts` field will be following:
> 
> ```ts
> @OneToMany(() => ---, (posts: Posts) => posts.posts)
> posts?: Posts[];
> ```
> 
> actually it should be:
> 
> ```ts
> @OneToMany(() => Post, (post: Post) => post.author)
> posts?: Post[];
> ```

Fixed and added tests. Now there's a convenient method `tests/model/fromSchemaString` for testing various combinations with inline schemas, so we can easily extend the tests as more bugs are spotted.

dzhelezov committed:
add isArray(), isScalar() helper methods

dzhelezov committed:
fix array types and add unit test

bedeho reviewed (approved):
Epic work, really nicely done. I am just glancing through here, can't provide any quality assurance.
